Signal/src/call/CallService.swift

```diff
@@ -495,58 +470,52 @@ protocol CallServiceObserver: class {
-            }
+        // BEWARE there are some unmerged changes around this that will likely cause a merge conflict.
+        // see: https://github.com/WhisperSystems/Signal-iOS/pull/2261/files
+        if let existingCall = self.call {
+            guard existingCall.remotePhoneNumber == thread.contactIdentifier(), existingCall.signalingId == callId else {
 
-            return
+                // TODO on iOS10+ we can use CallKit to swap calls rather than just returning busy immediately.
+                Logger.info("\(TAG) receivedCallOffer: \(newCall.identifiersForLogs) but we're already in call: \(existingCall.identifiersForLogs)")
+
+                handleLocalBusyCall(newCall, thread: thread)
+
+                if existingCall.remotePhoneNumber == newCall.remotePhoneNumber {
+                    // If we're receiving a new call offer from the user we think we have a call with, terminate
+                    // our current call to get back to a known good state.  If they call back, we'll be ready.
+                    //
+                    // TODO: Auto-accept this incoming call if our current call was either a) outgoing or
+                    // b) ever connected.  There will be a bit of complexity around making sure that two
+                    // parties that call each other at the same time end up connected.
+                    terminateCall()
+                }
+
+                return
+            }
         }
 
-        Logger.info("\(TAG) starting new call: \(newCall.identifiersForLogs)")
+        let incomingCall = self.call ?? newCall
+        Logger.info("\(TAG) starting new call: \(incomingCall.identifiersForLogs)")
 
-        self.call = newCall
+        self.call = incomingCall
 
         let backgroundTask = UIApplication.shared.beginBackgroundTask {
             let timeout = CallError.timeout(description: "background task time ran out before call connected.")
             DispatchQueue.main.async {
-                guard self.call == newCall else {
+                guard self.call == incomingCall else {
                     Logger.warn("\(self.TAG) ignoring obsolete call in \(#function)")
                     return
                 }
-                self.handleFailedCall(failedCall: newCall, error: timeout)
+                self.handleFailedCall(failedCall: incomingCall, error: timeout)
             }
         }
 
         let incomingCallPromise = firstly {
-            return getIceServers()
-        }.then { (iceServers: [RTCIceServer]) -> Promise<HardenedRTCSessionDescription> in
-            // FIXME for first time call recipients I think we'll see mic/camera permission requests here,
-            // even though, from the users perspective, no incoming call is yet visible.
-            guard self.call == newCall else {
-                throw CallError.obsoleteCall(description: "getIceServers() response for obsolete call")
+            return getPeerConnectionClientForIncomingCall(incomingCall)
```

Comment:
Because we can't guarantee message ordering, it's possible the callee would receive an important ICE update **before** receiving this CallOffer. Because both `handleCallOffer` and `handleRemoteAddedIceCandidate` need a peerConnectionClient to be in place,  I extracted the build-peer-connection logic behind a shared promise.